src/audio/module_adapter/module_adapter.c

```diff
@@ -867,6 +871,10 @@ static int module_adapter_audio_stream_type_copy(struct comp_dev *dev)
 	uint32_t num_input_buffers, num_output_buffers;
 	int ret, i = 0;
 
+	/* handle special case of HOST/DAI type components */
+	if (dev->ipc_config.type == SOF_COMP_HOST || dev->ipc_config.type == SOF_COMP_DAI)
+		return module_process_legacy(mod, NULL, 0, NULL, 0);
```

Comment:
@lyakh its not that the host/dai will process data with no src/sink buffers/ Its just that they're special in terms of what the source/sink buffers are. For the host playback case, the source buffer is the DMA buffer and similarly for the capture case, the sink buffer is the DMA buffer. So I completely skip the part with source/sink buffers and let the module itself handle it.

Reply:
hope in future, we can have a better way to handle all the cases instead of use "if", two disadvantages for this:
1. put more burden to last part of code execution.
2. code is not easy to understand.


like copier_copy(copier_process), we finally make it looks much more clear than initial state, I hope we can do similar on this too.